mne/epochs.py

```diff
@@ -2640,6 +2696,134 @@ def _get_epoch_from_raw(self, idx, verbose=None):
                                             self.reject_by_annotation)
         return data
 
+    @property
+    def raw_annotations(self):  # noqa: D102
+        return self._annotations
+
+    @verbose
+    def _set_annotations(self, annotations, on_missing='raise'):
+        """Setter for Epoch annotations from Raw.
+
+        This private function simply copies over Raw annotations, and
+        does not handle offsetting the times based on first_samp
+        or measurement dates, since that is expected to occur in
+        Raw set_annotations.
+
+        Parameters
+        ----------
+        annotations : instance of mne.Annotations | None
+            Annotations to set. If None, the annotations is defined
+            but empty.
+        %(on_missing_ch_names)s
+
+        Returns
+        -------
+        self : instance of Raw
+            The raw object with annotations.
+        """
+        meas_date = _handle_meas_date(self.info['meas_date'])
+        if annotations is None:
+            self._annotations = Annotations([], [], [], meas_date)
+        else:
+            _validate_type(annotations, Annotations, 'annotations')
+
+            if meas_date is None and annotations.orig_time is not None:
+                raise RuntimeError('Ambiguous operation. Setting an Annotation'
+                                   ' object with known ``orig_time`` to a raw'
+                                   ' object which has ``meas_date`` set to'
+                                   ' None is ambiguous. Please, either set a'
+                                   ' meaningful ``meas_date`` to the raw'
+                                   ' object; or set ``orig_time`` to None in'
+                                   ' which case the annotation onsets would be'
+                                   ' taken in reference to the first sample of'
+                                   ' the raw object.')
+            new_annotations = annotations.copy()
+            new_annotations._prune_ch_names(self.info, on_missing)
+            self._annotations = new_annotations
+
+        return self
+
+    def get_epoch_annotations(self):
+        """Return a list of raw annotations per epoch.
+
+        Returns
+        -------
+        epoch_annots : list
+            A list (with length equal to number of epochs) of annotations per
+            epoch. Each annotation is stored as (onset, duration, description),
+            where the onset is now relative to the epoch, rather then in
+            continuous time.
+        """
+        # check if annotations exist
+        if self.raw_annotations is None:
+            return self
+
+        events = self.events
+
+        # create a list of annotations for each epoch
+        epoch_annot_list = [[] for _ in range(len(events))]
+
+        for annot in self.raw_annotations:
```

Comment:
I wonder if it's possible to avoid a for-loop here?  `self.raw_annotations.onset` is a numpy array of all onsets, and similarly for `.duration`.  Seems like from that and `self.events[:, 0]`, `tmin` and `tmax`, and some combination of `np.searchsorted()` / `<` / `>` / bitwise `&` and `|` you could get there?

Reply:
Hmm not sure. I suppose this might be a performance issue if annotations is huge?

Not sure what you specifically have in mind here though. Would this remove just one for loop, or both for loops?